src/lib/io/sol-uart-impl-linux.c

```diff
@@ -172,72 +293,85 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
 SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
+    struct sol_blob *blob;
+    uint16_t i;
+
     SOL_NULL_CHECK(uart);
-    if (uart->async.rx_fd_handler)
-        sol_fd_del(uart->async.rx_fd_handler);
-    if (uart->async.tx_fd_handler)
-        sol_fd_del(uart->async.tx_fd_handler);
+
+    if (uart->in_use) {
+        uart->delete_me = true;
+        return;
+    }
+
+    uart->in_use = true;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i) {
+        if (uart->tx_cb)
+            uart->tx_cb((void *)uart->user_data, uart, blob, -ECANCELED);
+        sol_blob_unref(blob);
+    }
+
+    if (uart->rx.used)
+        uart->rx_cb((void *)uart->user_data, uart, &uart->rx);
+
+    if (uart->fd_handler)
+        sol_fd_del(uart->fd_handler);
+
+    if (uart->read_timeout)
+        sol_timeout_del(uart->read_timeout);
+
+    sol_ptr_vector_clear(&uart->pending_blobs);
+    sol_buffer_fini(&uart->rx);
     close(uart->fd);
     free(uart);
 }
 
-static void
-uart_tx_dispatch(struct sol_uart *uart, int status)
+SOL_API int
+sol_uart_write(struct sol_uart *uart, struct sol_blob *blob)
 {
-    uart->async.tx_fd_handler = NULL;
-    if (!uart->async.tx_cb)
-        return;
-    uart->async.tx_cb((void *)uart->async.tx_user_data, uart,
-        (uint8_t *)uart->async.tx_buffer, status);
-}
+    int r;
+    size_t total;
+    bool fd_created = false;
 
-static bool
-uart_tx_callback(void *data, int fd, uint32_t active_flags)
-{
-    struct sol_uart *uart = data;
-    int ret;
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
 
-    if (active_flags & FD_ERROR_FLAGS) {
-        SOL_ERR("Error flag was set on UART file descriptor %d.", fd);
-        ret = -1;
-        goto error;
-    }
+    if (uart->tx_size) {
+        r = sol_util_size_add(uart->pending_tx, blob->size, &total);
+        SOL_INT_CHECK(r, < 0, r);
```

Comment:
move the size_add outside of `if (uart->tx_size)` and always check for overflow, then you use `uart->pending_tx = total`